### PR TITLE
Port in torchified PPO from sb3_contrib

### PR DESCRIPTION
**Pull request by rhaps0dy**
I started by copying over the numpy code and then fixing whatever broke. This led to working code, unlike previous iterations. I also added the `PPORecurrent` algorithm to a bunch of existing tests, which is the same thing that `sb3_contrib` does.

This PR contains lots of Mypy type errors. I want to fix them when I make things generic over the hidden state, as in #4, and not now. If you check the CircleCI job you'll see that the tests themselves pass.

Other changes:
* Increased test parallelism in CircleCI to make tests less slow (there are some slow new tests)

### CODE REVIEW

**Comment by dan-pandori** [comment]
Can you add some lineage of what was copied from where? This is a big PR so it's going to take me a while to review, and understanding what is new code vs copied code would be helpful.

File: .circleci/config.yml [hunk 1]
```diff
@@ -71,7 +71,7 @@ jobs:
           password: "$GHCR_DOCKER_TOKEN"
     resource_class: medium
     working_directory: /workspace/third_party/stable-baselines3
-    parallelism: 16
+    parallelism: 24
```
Comment (dan-pandori) [inline .circleci/config.yml]: This seems irrelevant to this PR. Please either split it out into its own PR, or add it into the PR description.

File: stable_baselines3/common/buffers.py [hunk 2]
```diff
@@ -457,7 +457,7 @@ def compute_returns_and_advantage(self, last_values: th.Tensor, dones: th.Tensor
                 next_non_terminal = ~dones
                 next_values = last_values
             else:
-                next_non_terminal = 1.0 - self.episode_starts[step + 1]
+                next_non_terminal = ~self.episode_starts[step + 1]
```
Comment (dan-pandori) [inline stable_baselines3/common/buffers.py]: Why does this use bit negation now? The change from `1.0 - x` to `~x` seems like a semantically large one.

Comment (dan-pandori) [inline stable_baselines3/common/buffers.py]: `next_non_terminal` seems like a confusing name here. Thoughts on `is_next_non_terminal`?

`next_non_terminal` reads to me as referring to the next nonterminal value, but this variable seems to be referring to whether the following value to be iterated through is nonterminal.

**Review by dan-pandori — COMMENTED** [review]
Added some quick comments to the smaller files. Waiting to hear back on the sources of the larger files before reviewing them.

Thanks for the refactors in the tests to treating OnPolicy/OffPolicy algorithms separately rather than using sets, seems much cleaner now!

File: stable_baselines3/common/buffers.py (hunk 2 above)
Comment (rhaps0dy) [reply stable_baselines3/common/buffers.py]: It used to be that `episode_starts` was either a bool or a float. `1 - (bool tensor)` is supported in numpy but not in pytorch. It's now always a boolean tensor and gets negated.

Comment (dan-pandori) [reply stable_baselines3/common/buffers.py]: SG, 1-(bool) is yucky.

File: tests/test_buffers.py [hunk 3]
```diff
@@ -119,10 +136,17 @@ def test_device_buffer(replay_buffer_cls, device):
...
         ReplayBuffer: DummyEnv,
         DictReplayBuffer: DummyDictEnv,
+        RecurrentRolloutBuffer: DummyEnv,
+        RecurrentDictRolloutBuffer: DummyDictEnv,
     }[replay_buffer_cls]
     env = make_vec_env(env)
 
-    buffer = replay_buffer_cls(100, env.observation_space, env.action_space, device=device)
+    if replay_buffer_cls == RecurrentRolloutBuffer:
+        buffer = RecurrentRolloutBuffer(
```
Comment (dan-pandori) [inline tests/test_buffers.py:147]: Optional:
Here and below, `100` comes off as a bit of a magic number that it's hard to see why it was picked (and what it does). Passing all these in as kwargs would make it a little more readable.

If this 100 is the 'same' 100 that is used in line 153, it would be good to refactor them to use a shared constant.

**Review by dan-pandori — APPROVED** [review]
LGTM, thanks for splitting out the other parts!
